File: 백준/Silver/9372. 상근이의 여행/상근이의 여행.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/** 
하나의 연결요소로 만드는 적은 비용.

다 이어야함.
연결되는 두개 숫자도 정렬하고
그 다음 숫자리스트(비행기)도 정렬한 다음.
연결안돼있으면 유니온.하면서 cnt세줌
*/

int n, m; 

const int MX = 1002;
int par[MX];

vector<pair<int, int>> vec;

int find(int x){
    if(par[x] == x) return x;
    else return par[x] =find(par[x]);
}

void union_(int a, int b) {
    a = find(a);
    b = find(b);
    
    if(a == b) return;
    par[a] = b;
}

void solution(){
    cin >> n >> m;
    vec = {};
    for(int i = 1; i <= n; i++) par[i] = i;
    while(m--){
        int a, b;
        cin >> a >> b;
        if(a > b) vec.push_back({b,a});
        else vec.push_back({a,b});
    }
    
    sort(vec.begin(), vec.end());
    
    int cnt = 0;
    
    for(int i = 0; i < vec.size(); i++){
        int a = find(vec[i].first);
        int b = find(vec[i].second);
        
        if(a == b) continue;
        union_(a,b);
        cnt++;
    }
    
    cout << cnt << '\n';
}

int main()
{
    int t;
    cin >> t;
    while(t--){
        solution();
    }
    
    return 0;
}

File: 백준/Silver/9372. 상근이의 여행/상근이의 여행.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {
	static int par[] = new int[100020];
	static ArrayList<int[]> vec = new ArrayList<>();
	static BufferedReader br;
	static StringTokenizer st;
	static int V, E;
	static StringBuilder sb = new StringBuilder();

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));

		int T = Integer.parseInt(br.readLine());

		for (int t = 1; t <= T; t++) {
			st = new StringTokenizer(br.readLine());
			V = Integer.parseInt(st.nextToken());
			E = Integer.parseInt(st.nextToken());
			solution();
		}

		System.out.println(sb);
	}

	static void solution() throws Exception {
		init();
		int cnt = 0;
		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());

			if (find(a) != find(b)) {
				union(a, b);
				cnt++;
			}

		}

		sb.append(cnt).append("\n");
	}

	static void init() {
		for (int i = 0; i < 100020; i++) {
			par[i] = i;
		}
	}

	static int find(int a) {
		if (par[a] == a)
			return a;
		return par[a] = find(par[a]);
	}

	static void union(int a, int b) {
		a = find(a);
		b = find(b);
		if (a == b)
			return;
		par[b] = a;
	}
}